File: extscache/omni.kit.viewport.utility-1.1.2+69cbf6ad/omni/kit/viewport/utility/camera_state.py
# Copyright (c) 2022, NVIDIA CORPORATION.  All rights reserved.
#
# NVIDIA CORPORATION and its licensors retain all intellectual property
# and proprietary rights in and to this software, related documentation
# and any modifications thereto.  Any use, reproduction, disclosure or
# distribution of this software and related documentation without an express
# license agreement from NVIDIA CORPORATION is strictly prohibited.
#

__all__ = ["ViewportCameraState"]


from pxr import Gf, Sdf, Usd, UsdGeom


class ViewportCameraState:
    """A class that encapsulates the state of a camera within a viewport.

    It provides methods to query and manipulate the camera's position, target, and other properties, taking into account the stage's up axis and other USD-specific details.

    Args:
        camera_path: str
            The path to the camera in USD.
        viewport: The viewport associated with the camera state.
        time: Usd.TimeCode
            The time code for sampling the camera data."""

    def __init__(
        self, camera_path: str = None, viewport=None, time: Usd.TimeCode = None, **kwargs
    ):
        """Initializes the viewport camera state with a given camera path, viewport, and time

        Args:
            camera_path (str, optional): The path to the camera in USD. Defaults to the active viewport camera path if not provided.
            viewport (optional): The viewport to associate with the camera state. Defaults to the active viewport if not provided.
            time (Usd.TimeCode, optional): The time code for sampling the camera data. Defaults to Usd.TimeCode.Default() if not provided.

        Raises:
            RuntimeError: If no default or provided viewport is found."""
        if kwargs.get("force_legacy_api", None) != None:
            carb.log_error("force_legacy_api is deprecated arg to ViewportCameraState")

        if viewport is None:
            from omni.kit.viewport.utility import get_active_viewport

            viewport = get_active_viewport()
            if viewport is None:
                raise RuntimeError("No default or provided Viewport")

        self.__viewport_api = viewport
        self.__camera_path = str(camera_path if camera_path else viewport.camera_path)
        self.__time = Usd.TimeCode.Default() if time is None else time

    def get_world_camera_up(self, stage) -> Gf.Vec3d:
        """Returns the world space up direction for the camera based on the stage's up axis.

        Args:
            stage (Usd.Stage): The stage from which to retrieve the up axis information.

        Returns:
            Gf.Vec3d: The world space up direction vector for the camera."""
        up_axis = UsdGeom.GetStageUpAxis(stage) if stage else UsdGeom.Tokens.y
        if up_axis == UsdGeom.Tokens.y:
            return Gf.Vec3d(0, 1, 0)
        if up_axis == UsdGeom.Tokens.z:
            return Gf.Vec3d(0, 0, 1)
        if up_axis == UsdGeom.Tokens.x:
            return Gf.Vec3d(1, 0, 0)
        return Gf.Vec3d(0, 1, 0)

    @property
    def usd_camera(self) -> Usd.Prim:
        """Retrieves the USD camera primitive associated with the camera path.

        Returns:
            Usd.Prim: The USD camera primitive.

        Raises:
            RuntimeError: If the camera path does not correspond to a valid Usd.Prim or UsdGeom.Camera."""
        camera_prim = self.__viewport_api.stage.GetPrimAtPath(self.__camera_path)
        usd_camera = UsdGeom.Camera(camera_prim) if camera_prim else None
        if usd_camera:
            return usd_camera
        raise RuntimeError(f'"{self.__camera_path}" is not a valid Usd.Prim or UsdGeom.Camera')

    @property
    def position_world(self):
        """Gets the world position of the camera.

        Returns:
            Gf.Vec3d: The world position of the camera as a Gf.Vec3d vector."""
        return self.usd_camera.ComputeLocalToWorldTransform(self.__time).Transform(Gf.Vec3d(0, 0, 0))

    @property
    def target_world(self):
        """Gets the world space target position of the camera.

        Returns:
            Gf.Vec3d: The world space target position of the camera as a Gf.Vec3d vector."""
        local_coi = self.usd_camera.GetPrim().GetAttribute("omni:kit:centerOfInterest").Get(self.__time)
        return self.usd_camera.ComputeLocalToWorldTransform(self.__time).Transform(local_coi)

    def set_position_world(self, world_position: Gf.Vec3d, rotate: bool):
        """Sets the world position of the camera, with an option to rotate it to maintain its current orientation.

        Args:
            world_position (Gf.Vec3d): The new world position for the camera.
            rotate (bool): If True, the camera will be rotated to maintain its current orientation towards the center of interest.
        """
        usd_camera = self.usd_camera
        world_xform = usd_camera.ComputeLocalToWorldTransform(self.__time)
        parent_xform = usd_camera.ComputeParentToWorldTransform(self.__time)
        iparent_xform = parent_xform.GetInverse()
        initial_local_xform = world_xform * iparent_xform
        pos_in_parent = iparent_xform.Transform(world_position)
        if rotate:
            cam_prim = usd_camera.GetPrim()
            coi_attr = cam_prim.GetAttribute("omni:kit:centerOfInterest")
            prev_local_coi = coi_attr.Get(self.__time)
            coi_in_parent = iparent_xform.Transform(world_xform.Transform(prev_local_coi))
            cam_up = self.get_world_camera_up(cam_prim.GetStage())
            new_local_transform = Gf.Matrix4d(1).SetLookAt(pos_in_parent, coi_in_parent, cam_up).GetInverse()
        else:
            coi_attr, prev_local_coi = None, None
            new_local_transform = Gf.Matrix4d(initial_local_xform)

        new_local_transform = new_local_transform.SetTranslateOnly(pos_in_parent)

        import omni.kit.commands

        omni.kit.commands.create(
            "TransformPrimCommand",
            path=self.__camera_path,
            new_transform_matrix=new_local_transform,
            old_transform_matrix=initial_local_xform,
            time_code=self.__time,
            usd_context_name=self.__viewport_api.usd_context_name,
        ).do()
        if coi_attr and prev_local_coi:
            prev_world_coi = world_xform.Transform(prev_local_coi)
            new_local_coi = (new_local_transform * parent_xform).GetInverse().Transform(prev_world_coi)
            omni.kit.commands.create(
                "ChangePropertyCommand",
                prop_path=coi_attr.GetPath(),
                value=new_local_coi,
                prev=prev_local_coi,
                timecode=self.__time,
                usd_context_name=self.__viewport_api.usd_context_name,
                type_to_create_if_not_exist=Sdf.ValueTypeNames.Vector3d,
            ).do()

    def set_target_world(self, world_target: Gf.Vec3d, rotate: bool):
        """Sets the world target of the camera, with an option to rotate it. This can either move the camera to keep the same orientation and distance or just rotate the camera to look at the new target.

        Args:
            world_target (Gf.Vec3d): The new world target position for the camera.
            rotate (bool): If True, the camera will rotate to look at the new target. If False, the camera will move to maintain its orientation and distance relative to the target.
        """
        usd_camera = self.usd_camera
        world_xform = usd_camera.ComputeLocalToWorldTransform(self.__time)
        parent_xform = usd_camera.ComputeParentToWorldTransform(self.__time)
        iparent_xform = parent_xform.GetInverse()
        initial_local_xform = world_xform * iparent_xform

        cam_prim = usd_camera.GetPrim()
        coi_attr = cam_prim.GetAttribute("omni:kit:centerOfInterest")
        prev_local_coi = coi_attr.Get(self.__time)

        pos_in_parent = iparent_xform.Transform(initial_local_xform.Transform(Gf.Vec3d(0, 0, 0)))
        if rotate:
            # Rotate camera to look at new target, leaving it where it is
            cam_up = self.get_world_camera_up(cam_prim.GetStage())
            coi_in_parent = iparent_xform.Transform(world_target)
            new_local_transform = Gf.Matrix4d(1).SetLookAt(pos_in_parent, coi_in_parent, cam_up).GetInverse()
            new_local_coi = (new_local_transform * parent_xform).GetInverse().Transform(world_target)
        else:
            # Camera keeps orientation and distance relative to target
            # Calculate movement of center-of-interest in parent's space
            target_move = iparent_xform.Transform(world_target) - iparent_xform.Transform(
                world_xform.Transform(prev_local_coi)
            )
            # Copy the camera's local transform
            new_local_transform = Gf.Matrix4d(initial_local_xform)
            # And move it by the delta
            new_local_transform.SetTranslateOnly(pos_in_parent + target_move)

        import omni.kit.commands

        if rotate:
            omni.kit.commands.create(
                "ChangePropertyCommand",
                prop_path=coi_attr.GetPath(),
                value=new_local_coi,
                prev=prev_local_coi,
                timecode=self.__time,
                usd_context_name=self.__viewport_api.usd_context_name,
                type_to_create_if_not_exist=Sdf.ValueTypeNames.Vector3d,
            ).do()
        omni.kit.commands.create(
            "TransformPrimCommand",
            path=self.__camera_path,
            new_transform_matrix=new_local_transform,
            old_transform_matrix=initial_local_xform,
            time_code=self.__time,
            usd_context_name=self.__viewport_api.usd_context_name,
        ).do()

File: extscache/omni.graph.ui_nodes-1.50.5+107.3.0.lx64.r.cp311.u353/omni/graph/ui_nodes/ogn/nodes/CameraState.cpp
// SPDX-FileCopyrightText: Copyright (c) 2022-2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: LicenseRef-NvidiaProprietary
//
// NVIDIA CORPORATION, its affiliates and licensors retain all intellectual
// property and proprietary rights in and to this material, related
// documentation and any modifications thereto. Any use, reproduction,
// disclosure or distribution of this material and related documentation
// without an express license agreement from NVIDIA CORPORATION or
// its affiliates is strictly prohibited.

#include "CameraState.h"

#include <carb/settings/ISettings.h>

// clang-format off
#include <omni/graph/core/PreUsdInclude.h>
#include <pxr/usd/sdf/valueTypeName.h>
#include <omni/graph/core/PostUsdInclude.h>
// clang-format on

#include <omni/timeline/ITimeline.h>
#include <omni/usd/UsdUtils.h>


using namespace omni::graph::ui_nodes;

static const PXR_NS::TfToken kCenterOfInterest("omni:kit:centerOfInterest");

static bool checkPositionAndTarget(const PXR_NS::GfVec3d& position, const PXR_NS::GfVec3d& target)
{
    // If position and target are coincident, fail
    if ((position - target).GetLengthSq() <= std::numeric_limits<double>::epsilon())
    {
        return false;
    }
    return true;
}

static PXR_NS::GfVec3d getCameraUp(PXR_NS::UsdStageRefPtr stage)
{
    PXR_NS::TfToken upAxis = PXR_NS::UsdGeomGetStageUpAxis(stage);
    if (upAxis == PXR_NS::UsdGeomTokens->x)
    {
        return { 1, 0, 0 };
    }
    if (upAxis == PXR_NS::UsdGeomTokens->z)
    {
        return { 0, 0, 1 };
    }
    return { 0, 1, 0 };
}

CameraState::CameraState(PXR_NS::UsdGeomCamera camera, const PXR_NS::UsdTimeCode* time)
    : m_camera(std::move(camera)),
      m_timeCode(time ? *time :
                        omni::timeline::getTimeline()->getCurrentTime() *
                            m_camera.GetPrim().GetStage()->GetTimeCodesPerSecond())
{
}

void CameraState::getCameraPosition(carb::Double3& position) const
{
    PXR_NS::GfMatrix4d worldXform = m_camera.ComputeLocalToWorldTransform(m_timeCode).RemoveScaleShear();
    PXR_NS::GfVec3d worldPos = worldXform.Transform(PXR_NS::GfVec3d(0, 0, 0));
    position = { worldPos[0], worldPos[1], worldPos[2] };
}

static PXR_NS::UsdAttribute createCoiAttr(const PXR_NS::UsdPrim& cameraPrim)
{
    return cameraPrim.CreateAttribute(
        kCenterOfInterest, PXR_NS::SdfValueTypeNames->Vector3d, true, PXR_NS::SdfVariabilityUniform);
}

void CameraState::getCameraTarget(carb::Double3& target) const
{
    PXR_NS::GfVec3d localCenterOfInterest;
    PXR_NS::UsdAttribute coiAttr = m_camera.GetPrim().GetAttribute(kCenterOfInterest);
    if (!coiAttr || !coiAttr.Get(&localCenterOfInterest, m_timeCode))
    {
        localCenterOfInterest = { 0, 0, -1 };
    }

    PXR_NS::GfMatrix4d worldXform = m_camera.ComputeLocalToWorldTransform(m_timeCode).RemoveScaleShear();
    PXR_NS::GfVec3d worldCenterOfInterest = worldXform.Transform(localCenterOfInterest);
    target = { worldCenterOfInterest[0], worldCenterOfInterest[1], worldCenterOfInterest[2] };
}

bool CameraState::setCameraPosition(const carb::Double3& worldPosition, bool rotate)
{
    PXR_NS::GfMatrix4d worldXform = m_camera.ComputeLocalToWorldTransform(m_timeCode).RemoveScaleShear();
    PXR_NS::GfMatrix4d parentXform = m_camera.ComputeParentToWorldTransform(m_timeCode);
    PXR_NS::GfMatrix4d invParentXform = parentXform.GetInverse();
    PXR_NS::GfMatrix4d initialLocalXform = worldXform * invParentXform;
    PXR_NS::GfVec3d posInParent =
        invParentXform.Transform(PXR_NS::GfVec3d(worldPosition.x, worldPosition.y, worldPosition.z));
    PXR_NS::UsdPrim camPrim = m_camera.GetPrim();

    PXR_NS::UsdAttribute coiAttr;
    PXR_NS::GfVec3d prevLocalCenterOfInterest;
    PXR_NS::GfMatrix4d newLocalXform;
    if (rotate)
    {
        const PXR_NS::GfVec3d camUp = getCameraUp(camPrim.GetStage());

        coiAttr = camPrim.GetAttribute(kCenterOfInterest);
        if (!coiAttr || !coiAttr.Get(&prevLocalCenterOfInterest, m_timeCode))
        {
            prevLocalCenterOfInterest = { 0, 0, -1 };
            // Try to create the centerOfInterest attribute for set below
            coiAttr = createCoiAttr(camPrim);
        }

        PXR_NS::GfVec3d coiInParent = invParentXform.Transform(worldXform.Transform(prevLocalCenterOfInterest));
        if (!checkPositionAndTarget(posInParent, coiInParent))
        {
            return false;
        }
        newLocalXform = PXR_NS::GfMatrix4d(1).SetLookAt(posInParent, coiInParent, camUp).GetInverse();
    }
    else
    {
        newLocalXform = initialLocalXform;
    }

    newLocalXform.SetTranslateOnly(posInParent);
    omni::usd::UsdUtils::setLocalTransformMatrix(camPrim, newLocalXform, m_timeCode);

    if (coiAttr)
    {
        PXR_NS::GfVec3d prevWorldCOI = worldXform.Transform(prevLocalCenterOfInterest);
        PXR_NS::GfVec3d newLocalCOI = (newLocalXform * parentXform).GetInverse().Transform(prevWorldCOI);
        omni::usd::UsdUtils::setAttribute(coiAttr, newLocalCOI, m_timeCode);
    }

    return true;
}

bool CameraState::setCameraTarget(const carb::Double3& worldTarget, bool rotate)
{
    PXR_NS::UsdPrim camPrim = m_camera.GetPrim();

    PXR_NS::GfMatrix4d worldXform = m_camera.ComputeLocalToWorldTransform(m_timeCode).RemoveScaleShear();
    PXR_NS::GfMatrix4d parentXform = m_camera.ComputeParentToWorldTransform(m_timeCode);
    PXR_NS::GfMatrix4d invParentXform = parentXform.GetInverse();
    PXR_NS::GfMatrix4d initialLocalXform = worldXform * invParentXform;
    PXR_NS::GfVec3d gfWorldTarget(worldTarget.x, worldTarget.y, worldTarget.z);

    PXR_NS::GfVec3d prevLocalCenterOfInterest;
    PXR_NS::UsdAttribute coiAttr = camPrim.GetAttribute(kCenterOfInterest);
    if (!coiAttr || !coiAttr.Get(&prevLocalCenterOfInterest, m_timeCode))
    {
        prevLocalCenterOfInterest = { 0, 0, -1 };
        if (rotate)
        {
            // Try to create the centerOfInterest attribute for set below
            coiAttr = createCoiAttr(camPrim);
        }
    }

    PXR_NS::GfVec3d posInParent = invParentXform.Transform(initialLocalXform.Transform(PXR_NS::GfVec3d(0, 0, 0)));
    PXR_NS::GfMatrix4d newLocalXform;
    PXR_NS::GfVec3d newLocalCenterOfInterest;
    if (rotate)
    {
        // Rotate camera to look at new target, leaving it where it is
        PXR_NS::GfVec3d camUp = getCameraUp(camPrim.GetStage());
        PXR_NS::GfVec3d coiInParent = invParentXform.Transform(gfWorldTarget);
        if (!checkPositionAndTarget(posInParent, coiInParent))
        {
            return false;
        }
        newLocalXform = PXR_NS::GfMatrix4d(1).SetLookAt(posInParent, coiInParent, camUp).GetInverse();
        newLocalCenterOfInterest = (newLocalXform * parentXform).GetInverse().Transform(gfWorldTarget);
    }
    else
    {
        // Camera keeps orientation and distance relative to target
        // Calculate movement of center-of-interest in parent's space
        PXR_NS::GfVec3d targetMove = invParentXform.Transform(gfWorldTarget) -
                                     invParentXform.Transform(worldXform.Transform(prevLocalCenterOfInterest));
        // Copy the camera's local transform
        newLocalXform = initialLocalXform;
        // And move it by the delta
        newLocalXform.SetTranslateOnly(posInParent + targetMove);
    }

    if (rotate)
    {
        omni::usd::UsdUtils::setAttribute(coiAttr, newLocalCenterOfInterest, m_timeCode);
    }
    omni::usd::UsdUtils::setLocalTransformMatrix(camPrim, newLocalXform, m_timeCode);

    return true;
}
